clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -566,18 +567,18 @@ public ConsumerGroupHeartbeatRequestData buildRequestData() {
 
             // ClientAssignors - not supported yet
 
-            // TopicPartitions - only sent if it has changed since the last heartbeat. Note that
-            // the string consists of just the topic ID and the partitions. When an assignment is
-            // received, we might not yet know the topic name, and then it is learnt subsequently
-            // by a metadata update.
-            TreeSet<String> assignedPartitions = membershipManager.currentAssignment().entrySet().stream()
-                .map(entry -> entry.getKey() + "-" + entry.getValue())
-                .collect(Collectors.toCollection(TreeSet::new));
-            if (!assignedPartitions.equals(sentFields.topicPartitions)) {
+            // TopicPartitions - sent with the first heartbeat after a new assignment from the server was
+            // reconciled. This is ensured by resending the topic partitions whenever the local assignment,
+            // including its local epoch is changed (although the local epoch is not sent in the heartbeat).
+            LocalAssignment local = membershipManager.currentAssignment();
+            if (local == null) {
+                data.setTopicPartitions(Collections.emptyList());
+                sentFields.topicPartitions = null;
+            } else if (!local.equals(sentFields.topicPartitions)) {
```

Comment:
Don't we need to also take the assignment epoch into consideration here? In other words, should we store the LocalAssignment in `sentFields` and use it to do the comparison?

Reply:
That's what is being done. Maybe I should have renamed the field `sentFields.topicPartitions` to `sentFields.localAssignment`, but the type is now `LocalAssignment` and the `equals` compares the local epoch as well.

Reply:
Renamed the field